File: 2024/day1/pt1.py
import timeit
 
def run():
    f = open("input.txt", "r")
    data = f.read()
    lines = data.split("\n")
    delim = "   "
    left = []
    right = []
    for x in lines:
        br = x.split(delim)
        left.append(br[0])
        right.append(br[1])
 
    #fixing order
    left.sort()
    right.sort()
 
    loop = len(left)
    distance = 0
    for x in range(0, loop):
        l = int(left[x])
        r = int(right[x])
        distance += abs(l-r)
 
    print("distance: " + str(distance))
 
if __name__ == "__main__":
    timing = timeit.timeit(run, number=1)
    print("Execution time: " + str(timing) + " seconds")

#Execution time: 0.0012259000213816762 seconds

File: 2024/day2/pt1/pt1.go
package main

import (
	"io"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

func main() {
	st := time.Now()
	file, err := os.Open("../input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	data, err := io.ReadAll(file)
	if err != nil {
		panic(err)
	}

	dataStr := string(data)
	lines := strings.Split(dataStr, "\r\n") //windows newline
	safeCount := 0
	for _, l := range lines {
		ld := strings.Fields(l) //cant Split() on whitespace, using Fields will split on whitespace
		last := 0
		increasing := false
		decreasing := false
		safe := true

		for idx, d := range ld {
			nmbr, _ := strconv.Atoi(d)
			if idx == 0 {
				last = nmbr
				continue
			}

			if last < nmbr {
				increasing = true
			} else {
				decreasing = true
			}

			if last == nmbr {
				safe = false
				break
			}

			if math.Abs(float64(last-nmbr)) > 3 {
				safe = false
				break
			}
			last = nmbr
		}

		if increasing && decreasing {
			safe = false
			increasing = false
			decreasing = false
		}

		if safe {
			safeCount++
		}
	}

	log.Printf("%d safe reports\n", safeCount)
	log.Printf("Execution time: %s", time.Since(st).String())
}

//Execution time: 510.9µs
